### PR TITLE
feat: Subscription Service for Auto-Swap

### PR DESCRIPTION
**Pull request by Gerson2102**
## Issue:
Solves #2 

## Changes
Adding the subscription file handling all the logic related to the requirements.
Adding new types
Adding new route

## Notes
I tested locally and is working. But I'm not 100% sure if this is what you guys want. Is my first time doing something like this.

### CODE REVIEW

**Comment by Abeeujah** [comment]
Hi @Gerson2102 Great job so far, Could you add tests, and also run `cargo fmt` before you push, this should fix your CI build issue.

**Comment by Gerson2102** [comment]
AIght! @Abeeujah currently working on the tests.

**Comment by Abeeujah** [comment]
Hi @Gerson2102 can you please fix CI issue?

**Comment by Gerson2102** [comment]
@Abeeujah can you run the the CI again to see what happens now please?

**Comment by sprtd** [comment]
@Gerson2102 Kindly run `cargo fmt -- --check` locally; then fix the fmt error

File: tests/api/subscription.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        .method("POST")
+        .uri("/subscriptions")
+        .header(CONTENT_TYPE, "application/json")
+        .body(Body::from(serde_json::to_string(&payload).unwrap()))
+        .unwrap();
+
+    let resp = app.request(req).await;
+    assert_eq!(resp.status(), StatusCode::OK);
+}
+
```
Comment (Abeeujah) [inline tests/api/subscription.rs:57]: Hi, this test is failing due to you're passing an asynchronous operation into a synchronous operation, the order they complete is up to the workers that run the asynchronous operation, what I think you should test for is the `from_token` is stored correctly with the `percentage`
```rust
for (token, percentage) in payload.from_token.iter().zip(payload.percentage.iter()) {
        sqlx::query!(
            r#"
            INSERT INTO swap_subscription_from_token 
            (wallet_address, from_token, percentage)
            VALUES ($1, $2, $3)
            "#,
            payload.wallet_address,
            token,
            percentage,
        )
        .execute(&mut *tx)
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
    }
```
so for a retrieved record
```rust
sqlx::query!(
        r#"
        SELECT wallet_address, from_token, percentage 
        FROM swap_subscription_from_token
        WHERE wallet_address = $1 AND from_token = $2
        "#,
        wallet_address,
        from_token,
    )
    ```
    assert that the from_token percent is set properly.

**Review by Abeeujah — CHANGES_REQUESTED** [review]

File: src/http/subscription.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+use crate::AppState;
+
+pub async fn create_subscription(
+    State(state): State<AppState>,
+    Json(payload): Json<CreateSubscriptionRequest>,
+) -> Result<Json<CreateSubscriptionResponse>, StatusCode> {
+    if payload.from_token.len() != payload.percentage.len() {
+        return Err(StatusCode::BAD_REQUEST);
+    }
+
```
[the remark this answers is not in the document]
Comment (Gerson2102) [reply src/http/subscription.rs:14]: Got it

**Comment by Gerson2102** [comment]
I dont know how to make the tests working from the first time that we run `cargo test`. @Abeeujah

**Pull request merged**